Software/grab/MacOSGrabberBase.mm

```diff
@@ -0,0 +1,229 @@
+			const int x2 = displayRect.size.width  + x1 - 1;
+			const int y2 = displayRect.size.height + y1 - 1;
+			screenInfo.rect.setCoords(x1, y1, x2, y2);
+			screenInfo.handle = reinterpret_cast<void *>(display);
+
+			return true;
+		}
+	}
+	
+	return false;
+}
+
+double MacOSGrabberBase::getDisplayScalingRatio(const CGDirectDisplayID display)
+{
+	CGDisplayModeRef mode = CGDisplayCopyDisplayMode(display);
+	if (mode == NULL)
+		return 1.0;
+	const double scaledWidth = static_cast<double>(CGDisplayModeGetWidth(mode));
+	const double realWidth = static_cast<double>(CGDisplayModeGetPixelWidth(mode));
+	CGDisplayModeRelease(mode);
+	return (realWidth / scaledWidth);
+}
+
+double MacOSGrabberBase::getDisplayRefreshRate(const CGDirectDisplayID display)
+{
+	CGDisplayModeRef mode = CGDisplayCopyDisplayMode(display);
+	if (mode == NULL)
+		return 0.0;
+	const double hz = CGDisplayModeGetRefreshRate(mode);
+	CGDisplayModeRelease(mode);
+	return hz;
+}
+
+QList<ScreenInfo>* MacOSGrabberBase::screensWithWidgets(
+	QList<ScreenInfo>* result,
+	const QList<GrabWidget *> &grabWidgets)
+{
+	CGDirectDisplayID displays[kMaxDisplaysCount];
+	uint32_t displayCount = 0;
+
+	CGError err = CGGetActiveDisplayList(kMaxDisplaysCount, displays, &displayCount);
+
+	result->clear();
+	if (err == kCGErrorSuccess) {
+		for (unsigned int i = 0; i < displayCount; ++i) {
+			ScreenInfo screenInfo;
+			if (getScreenInfoFromRect(displays[i], grabWidgets, screenInfo))
+				result->append(screenInfo);
+		}
+
+	} else {
+		qCritical() << "couldn't get active displays, error code " << QString::number(err, 16);
+	}
+	return result;
+}
+
+bool MacOSGrabberBase::reallocate(const QList<ScreenInfo> &screens)
+{
+	freeScreens();
+	for (const ScreenInfo& screen : screens) {
+		GrabbedScreen grabScreen;
+		if (!allocateScreenBuffer(screen, grabScreen)) {
+			qCritical() << "couldn't allocate image buffer";
+			freeScreens();
+			return false;
+		}
+		_screensWithWidgets.append(grabScreen);
+
+	}
+	return true;
+}
+
+void MacOSGrabberBase::freeScreenImageData(GrabbedScreen& screen)
+{
+	screen.imgData = nullptr;
+	if (screen.associatedData)
+		delete ((MacOSScreenData*)screen.associatedData);
+	screen.associatedData = nullptr;
+}
+
+GrabResult MacOSGrabberBase::grabScreens()
+{
+#ifndef QT_NO_DEBUG
+	static unsigned long _count = 0;
+	_count += m_timer->interval();
+#endif
+
+	for (GrabbedScreen& grabScreen : _screensWithWidgets)
+	{
+		freeScreenImageData(grabScreen);
+		CGDirectDisplayID display = static_cast<CGDirectDisplayID>(reinterpret_cast<intptr_t>(grabScreen.screenInfo.handle));
+		
+		const GrabResult result = grabDisplay(display, grabScreen);
+
+		if (result == GrabResultError) {
+			qCritical() << Q_FUNC_INFO << "grabDisplay failed";
+			return GrabResultError;
+		} else if (result == GrabResultFrameNotReady)
+			return GrabResultFrameNotReady;
+
+#ifndef QT_NO_DEBUG
+		if (_count > 20000) // save every 20sec
+			saveGrabbedScreenToBMP(grabScreen);
```

Comment:
This shouldn't happen by default, not even in debug builds. Please have it disabled by default.

Reply:
what do you mean by part of the log? If you want, you can make it dependent on the log level

Reply:
you'll want to switch ifndef to ifdef ;)

Reply:
Well, along with the usual log text stuff I used this as a visual log.
>  If you want, you can make it dependent on the log level

Oh yeah, how can I link to it to `DEBUG_HIGH_LEVEL` for example?

On a side note: at some point I had it in a QPixmax preview under virtual leds which was cool (we can dump our RGB frames there from any platform), but I thought it was too much of a hassle to do properly without it being useful to the end user.

Reply:
yes exactly. Since it's a developer thing anyway I'm fine with what you have now, just flip the ifndefs to ifdefs

Reply:
alright, pushed.